### PR TITLE
feat: allow writing compressed files

### PR DESCRIPTION
**Pull request by radioflash**
This fixes the original effort from fguevaravas (#77),
also adding a testcase for compression effectiveness,
correcting some minor whitespace issues, and
adding an example to README.md.

### CODE REVIEW

**Comment by radioflash** [comment]
There is still room for improvement:

Reference file (MATLAB 5.0): 5MiB
Julia compressed: 18MiB
Julia uncompressed: 215MiB

Interestingly, the reference file contains no uncompressed strings whatsoever, while the compressed Julia version keeps basically every fieldname as plaintext (my example stores a single Dict: String => Array{Float64, 2}).

But this is currently good enough for me, and it already helps a TON.

File: src/MAT.jl [hunk 1]
```diff
@@ -85,9 +85,10 @@ function matopen(filename::AbstractString, rd::Bool, wr::Bool, cr::Bool, tr::Boo
 end
 
 function matopen(fname::AbstractString, mode::AbstractString)
-    mode == "r"  ? matopen(fname, true , false, false, false, false) :
-    mode == "r+" ? matopen(fname, true , true , false, false, false) :
-    mode == "w"  ? matopen(fname, false, true , true , true , false) :
+    mode == "r"  ? matopen(fname, true , false, false, false, false, false) :
+    mode == "r+" ? matopen(fname, true , true , false, false, false, false) :
+    mode == "w"  ? matopen(fname, false, true , true , true , false, false) :
+    mode == "wz" ? matopen(fname, false, true , true , true , false, true ) :
```
Comment (timholy) [inline src/MAT.jl]: Is this standard? Should be documented in any case.

Seems like `z` might be relevant for `r+`? If so, you could check `compress = endswith(mode, 'z')` and if true strip it off and then pass `compress` as the third argument.

File: src/MAT_HDF5.jl [hunk 2]
```diff
@@ -365,7 +374,7 @@ function m_write(mfile::MatlabHDF5File, parent::HDF5Parent, name::String, data::
         m_writeempty(parent, name, data)
         return
     end
-    dset = m_writearray(parent, name, toarray(data))
+    dset = m_writearray(parent, name, toarray(data),mfile.compress)
```
Comment (timholy) [inline src/MAT_HDF5.jl]:
```suggestion
    dset = m_writearray(parent, name, toarray(data), mfile.compress)
```

File: src/MAT_HDF5.jl [hunk 3]
```diff
@@ -101,7 +102,7 @@ function matopen(filename::AbstractString, rd::Bool, wr::Bool, cr::Bool, tr::Boo
         writeheader = false
     end
     close(pa)
-    fid = MatlabHDF5File(HDF5File(f, filename), true, writeheader, 0)
+    fid = MatlabHDF5File(HDF5File(f, filename), true, writeheader, 0,compress)
```
Comment (timholy) [inline src/MAT_HDF5.jl]:
```suggestion
    fid = MatlabHDF5File(HDF5File(f, filename), true, writeheader, 0, compress)
```

File: src/MAT.jl (hunk 1 above)
Comment (radioflash) [reply src/MAT.jl]: > Is this standard?
I took this from fguevaravas original PR tbh.

Maybe a keyword argument like `matopen(...; compress = true)` both for matopen and matwrite would be best?

Would you prefer keyword argument or "z" options for the mode-string?

Comment (timholy) [reply src/MAT.jl]: If the `wz` isn't standard, then I think the keyword might be best.

**Comment by timholy** [comment]
Looks like there's a merge conflict, probably from me merging the bit array test. Is your git-fu up to resolving it? If not, don't worry about it, I'll fix it.

File: src/MAT.jl (hunk 1 above)
Comment (radioflash) [reply src/MAT.jl]: On it!

**Comment by timholy** [comment]
Thanks @radioflash and @fguevaravas!

**Pull request merged**